File: scripts/generate_metadata.py
#!/usr/bin/env python3
import json
import os
from datetime import datetime

import boto3

# =========================
# CONFIG
# =========================

R2_ACCOUNT_ID = os.getenv("R2_ACCOUNT_ID", "")
R2_ACCESS_KEY_ID = os.getenv("R2_ACCESS_KEY_ID", "")
R2_SECRET_ACCESS_KEY = os.getenv("R2_SECRET_ACCESS_KEY", "")

if not R2_ACCOUNT_ID or not R2_ACCESS_KEY_ID or not R2_SECRET_ACCESS_KEY:
    raise SystemExit("Missing R2_ACCOUNT_ID, R2_ACCESS_KEY_ID, or R2_SECRET_ACCESS_KEY in environment.")

BUCKET_NAME = "kztg"

PREFIX = ""

PUBLIC_BASE_URL = "https://pub-8b5cdf2f2aed4a6e832dd72430dfacc1.r2.dev"

OUTPUT_FILE = "js/metadata.js"

# =========================
# R2 CLIENT
# =========================

s3 = boto3.client(
    "s3",
    endpoint_url=f"https://{R2_ACCOUNT_ID}.eu.r2.cloudflarestorage.com",
    aws_access_key_id=R2_ACCESS_KEY_ID,
    aws_secret_access_key=R2_SECRET_ACCESS_KEY,
    region_name="auto",
)

def public_url(key: str) -> str:
    return f"{PUBLIC_BASE_URL}/{key}"


# =========================
# LOAD OBJECTS
# =========================

sermons = {}

audio_priority = {
    ".opus": 0,
    ".mp3": 1,
    ".m4a": 2,
    ".aac": 3,
    ".ogg": 4,
    ".wav": 5,
}
cover_priority = {
    ".webp": 0,
    ".jpg": 1,
    ".jpeg": 2,
    ".png": 3,
}

continuation_token = None

while True:
    kwargs = {
        "Bucket": BUCKET_NAME,
        "Prefix": PREFIX,
    }

    if continuation_token:
        kwargs["ContinuationToken"] = continuation_token

    response = s3.list_objects_v2(**kwargs)

    for obj in response.get("Contents", []):
        key = obj["Key"]

        parts = key.split("/")

        if len(parts) != 3:
            continue

        _, sermon_id, filename = parts


        sermons.setdefault(sermon_id, {
            "id": sermon_id,
            "date": "",
        })
        sermons[sermon_id].setdefault("_audio_rank", None)
        sermons[sermon_id].setdefault("_cover_rank", None)

        if filename == "metadata.json":
            obj_data = s3.get_object(Bucket=BUCKET_NAME, Key=key)
            meta = json.loads(obj_data["Body"].read())

            if filename == "manifest.json":
                sermons[sermon_id]["date"] = meta.get("date", "")
                sermons[sermon_id]["preacher"] = meta.get("preacher", "")
                sermons[sermon_id]["title"] = meta.get("title", "")
            else:
                if not sermons[sermon_id].get("date"):
                    sermons[sermon_id]["date"] = meta.get("date", "")
                if not sermons[sermon_id].get("preacher"):
                    sermons[sermon_id]["preacher"] = meta.get("preacher", "")
                if not sermons[sermon_id].get("title"):
                    sermons[sermon_id]["title"] = meta.get("title", "")

        _, ext = os.path.splitext(filename)
        ext = ext.lower()

        if ext in audio_priority:
            current_rank = sermons[sermon_id]["_audio_rank"]
            new_rank = audio_priority[ext]
            if current_rank is None or new_rank < current_rank:
                sermons[sermon_id]["audio"] = public_url(key)
                sermons[sermon_id]["_audio_rank"] = new_rank

        if ext in cover_priority:
            current_rank = sermons[sermon_id]["_cover_rank"]
            new_rank = cover_priority[ext]
            if current_rank is None or new_rank < current_rank:
                sermons[sermon_id]["cover"] = public_url(key)
                sermons[sermon_id]["_cover_rank"] = new_rank

    if not response.get("IsTruncated"):
        break

    continuation_token = response.get("NextContinuationToken")

# =========================
# FILTER + SORT
# =========================

items = []

for item in sermons.values():
    if "audio" not in item:
        continue

    item.pop("_audio_rank", None)
    item.pop("_cover_rank", None)
    items.append(item)

def sort_key(entry):
    return (
        entry.get("date", ""),
        entry.get("id", ""),
    )

items.sort(key=sort_key, reverse=True)

# =========================
# FINAL JSON
# =========================

metadata = {
    "generated_at": datetime.now().isoformat() + "Z",
    "count": len(items),
    "items": items,
}

# =========================
# WRITE FILE
# =========================

with open(OUTPUT_FILE, "w", encoding="utf-8") as f:
    f.write("window.SERMONS=")
    json.dump(items, f, separators=(",", ":"))
    f.write(";")

print(f"Generated {OUTPUT_FILE}")
print(f"Items: {len(items)}")

File: scripts/generate_metadata.js
#!/usr/bin/env node
/**
 * Generates js/metadata.js by listing R2 (audio + metadata.json + covers).
 * Covers are downloaded from R2, converted to webp, and saved to dist/images/covers/.
 * Re-runs skip already-downloaded covers.
 */
import { S3Client, ListObjectsV2Command, GetObjectCommand } from "@aws-sdk/client-s3";
import sharp from "sharp";
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";
import { config as dotenvConfig } from "dotenv";

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
dotenvConfig({ path: path.join(ROOT, ".env") });

// =========================
// CONFIG
// =========================

const R2_ACCOUNT_ID        = process.env.R2_ACCOUNT_ID ?? "";
const R2_ACCESS_KEY_ID     = process.env.R2_ACCESS_KEY_ID ?? "";
const R2_SECRET_ACCESS_KEY = process.env.R2_SECRET_ACCESS_KEY ?? "";
const R2_ENDPOINT          = process.env.R2_ENDPOINT
  ?? `https://${R2_ACCOUNT_ID}.eu.r2.cloudflarestorage.com`;
const BUCKET_NAME          = process.env.R2_BUCKET ?? "kztg";
const PUBLIC_BASE_URL      = (
  process.env.R2_PUBLIC_BASE_URL
  ?? "https://pub-8b5cdf2f2aed4a6e832dd72430dfacc1.r2.dev"
).replace(/\/$/, "");

if (!R2_ACCOUNT_ID || !R2_ACCESS_KEY_ID || !R2_SECRET_ACCESS_KEY) {
  console.error("Missing R2_ACCOUNT_ID, R2_ACCESS_KEY_ID or R2_SECRET_ACCESS_KEY.");
  process.exit(1);
}

const OUTPUT_FILE  = path.join(ROOT, "js/metadata.js");
const COVERS_DIR   = path.join(ROOT, "dist/images/covers");
const COVERS_MAX_W = 640;

// =========================
// R2 CLIENT
// =========================

const s3 = new S3Client({
  endpoint: R2_ENDPOINT,
  region: "auto",
  credentials: { accessKeyId: R2_ACCESS_KEY_ID, secretAccessKey: R2_SECRET_ACCESS_KEY },
});

function publicUrl(key) {
  return `${PUBLIC_BASE_URL}/${key}`;
}

// =========================
// COVER: download from R2 → dist/images/covers/<id>.webp
// =========================

async function fetchCover(r2Key, sermonId) {
  fs.mkdirSync(COVERS_DIR, { recursive: true });
  const outPath = path.join(COVERS_DIR, `${sermonId}.webp`);

  if (fs.existsSync(outPath)) return `/images/covers/${sermonId}.webp`;

  try {
    const res = await s3.send(new GetObjectCommand({ Bucket: BUCKET_NAME, Key: r2Key }));
    const chunks = [];
    for await (const c of res.Body) chunks.push(c);
    const buffer = Buffer.concat(chunks);

    const img = sharp(buffer);
    const meta = await img.metadata();
    const pipeline = meta.width > COVERS_MAX_W
      ? img.resize(COVERS_MAX_W, null, { fit: "inside", kernel: "lanczos3" })
      : img;

    await pipeline.webp({ quality: 82, effort: 4 }).toFile(outPath);
    console.log(`  cover saved: dist/images/covers/${sermonId}.webp (${meta.width}×${meta.height})`);
    return `/images/covers/${sermonId}.webp`;
  } catch (err) {
    console.warn(`  WARNING: could not fetch cover for ${sermonId}: ${err.message}`);
    return publicUrl(r2Key);
  }
}

// =========================
// LIST R2
// =========================

const audioPriority = { ".opus": 0, ".mp3": 1, ".m4a": 2, ".aac": 3, ".ogg": 4, ".wav": 5 };
const coverPriority = { ".webp": 0, ".jpg": 1, ".jpeg": 2, ".png": 3 };
const sermons = {};
let token;

do {
  const res = await s3.send(new ListObjectsV2Command({
    Bucket: BUCKET_NAME,
    ...(token ? { ContinuationToken: token } : {}),
  }));

  for (const obj of res.Contents ?? []) {
    const parts = obj.Key.split("/");
    if (parts.length !== 3 || parts[0] !== "sermons") continue;

    const [, id, filename] = parts;
    const ext = path.extname(filename).toLowerCase();

    sermons[id] ??= { id, date: "", _audioRank: null, _coverRank: null };

    if (filename === "metadata.json") {
      const r = await s3.send(new GetObjectCommand({ Bucket: BUCKET_NAME, Key: obj.Key }));
      const chunks = [];
      for await (const c of r.Body) chunks.push(c);
      const meta = JSON.parse(Buffer.concat(chunks).toString("utf8"));
      if (!sermons[id].date)     sermons[id].date     = meta.date     ?? "";
      if (!sermons[id].preacher) sermons[id].preacher = meta.preacher ?? "";
      if (!sermons[id].title)    sermons[id].title    = meta.title    ?? "";
    }

    if (ext in audioPriority) {
      const rank = audioPriority[ext];
      if (sermons[id]._audioRank === null || rank < sermons[id]._audioRank) {
        sermons[id].audio = publicUrl(obj.Key);
        sermons[id]._audioRank = rank;
      }
    }

    if (ext in coverPriority) {
      const rank = coverPriority[ext];
      if (sermons[id]._coverRank === null || rank < sermons[id]._coverRank) {
        sermons[id]._coverKey = obj.Key;
        sermons[id]._coverRank = rank;
      }
    }
  }

  token = res.IsTruncated ? res.NextContinuationToken : undefined;
} while (token);

// =========================
// BUILD ITEMS
// =========================

const items = [];

for (const sermon of Object.values(sermons)) {
  if (!sermon.audio) continue;

  const coverKey = sermon._coverKey;
  if (coverKey) sermon.cover = await fetchCover(coverKey, sermon.id);

  delete sermon._audioRank;
  delete sermon._coverRank;
  delete sermon._coverKey;

  items.push(sermon);
}

items.sort((a, b) =>
  a.date < b.date ? 1 : a.date > b.date ? -1 : a.id < b.id ? 1 : a.id > b.id ? -1 : 0
);

// =========================
// WRITE
// =========================

fs.writeFileSync(OUTPUT_FILE, `window.SERMONS=${JSON.stringify(items)};`, "utf8");
console.log(`\nGenerated ${OUTPUT_FILE}`);
console.log(`Items: ${items.length}`);

